src/irmin-pack/ext.ml

```diff
@@ -187,12 +183,64 @@ module Maker (Config : Conf.S) = struct
                   Lwt.fail e
               | e -> Lwt.fail e)
 
-        (** Stores share instances in memory, one sync is enough. *)
-        let sync t = Contents.CA.sync (contents_t t)
+        (** For crash consistency and SWMR consistency, the order in which we
+            [flush], [sync] and [clear] the various files should carefuly chosen
+            in order to avoid introducing dangling references from file to file.
+
+            There are 4 types of files (called components below), the ones in
+            index, the dict, the pack file and the branch one. Here are the
+            relations between them:
+
+            - the branch store references commit hashes in index,
+            - index references pack entries,
+            - the pack file references dict entries and
+            - the dict references nothing.
+
+            In case of crash (or an ro instance that concurrently syncs), the
+            [clear] the order should be: branch, index, pack and dict.
+
+            In case an rw instance concurrently flushes, the [sync] the order
+            should be: branch, index, pack and dict.
```

Comment:
I don't understand why `sync` has to be done in reverse order. Why is this different from clear and flush?

Reply:
Actually, that's a very good point, read-only sync should be done in reverse order and I believe we are not doing this properly at the moment.

Reply:
It is `flush` that is done in the opposite order than `clear` and `sync`.

I can attempt another explanation. Consider this dependency graph:
```
(top)
branch
  |
  v
index
  |
  v
pack
  |
  v
dict
(bot)
```

### crash during `flush`
`flush` top to bot (not this PR): peer2 gets dangling dict references
```
RW: flush branches/index/pack  <crash>    
RO:                                     sync all
```

`flush` bot to top (this PR). peer2 gets __no__ gandling dict references
```
RW: flush dict  <crash>   
RO:                      sync all
```


### concurrent sync (rw hangs) during `flush`
`flush` top to bot (not this PR): RO gets dangling dict references
```
RW: flush branches/index/pack            flush dict
RO:                            sync all
```

`flush` bot to top (this PR): RO gets __no__ gangling dict references
```
RW: flush dict            flush index/pack/branches
RO:             sync all
```


### concurrent sync (ro hangs) during `flush`
`sync` bot to top (not this PR): RO gets dangling dict references.
```
RW:            flush all
RO: sync dict             sync index/pack/branches
```

`sync` top to bot (this PR): RO gets __no__ dangling dict references.
```
RW:                           flush all
RO: sync branches/index/pack             sync dict                    
```

### crash during `clear`
`clear` bot to top (not this PR). peer2 gets gandling dict references
```
RW: clear dict  <crash>
RO:                      sync all
```

`clear` top to bot (this PR): peer2 gets __no__ dangling dict references
```
RW: clear branches/index/pack  <crash>
RO:                                     sync all
```


### concurrent sync (rw hangs) during `clear`
`clear` bot to top (not this PR): RO gets gangling dict references
```
RW: clear dict            clear pack/index/branches
RO:             sync all
```

`clear` top to bot (this PR): RO gets __no__ dangling dict references
```
RW: clear branches/index/pack            clear dict
RO:                            sync all
```


### concurrent sync (ro hangs) during `clear` :warning: :warning:
`sync` bot to top (not this PR): RO gets __no__ dangling dict references :warning: :warning:
```
RW:            clear all
RO: sync dict             sync index/pack/branches
```

`sync` top to bot (this PR): RO gets dangling dict references :warning: :warning:

```
RW:                           clear all
RO: sync branches/index/pack             sync dict
```

Reply:
I now believe that `sync` can't be consistent with both `clear` and `flush`

Reply:
> It is flush that is done in the opposite order than clear and sync

this doesn't sounds right: clear and flush are write operations, they should be done in the the same order. Sync is a read operation and should be done in the reverse order to be consistent.

Reply:
`clear` isn't a regular write operation, it is the only operation that removes data from the store. If a writing and reading should have opposite orders, so does adding and removing data.

Changing clear to use the same order as `flush` wouldn't change the fact that `sync` can't be consistent with both `clear` and `flush` at the same time.

Reply:
Thanks @Ngoguey42 for this analysis, I agree that `sync` cannot be consistent with both `clear` and `flush`. 

But we don't support `clear` anymore, the only place is still used is when opening a RW instance with `fresh:true` while a RO instance is running (which is anyway not very well supported https://github.com/mirage/irmin/pull/955).  So I propose we move the comments about `clear` in the pack_store (where it is used), and not deal with the `clear`/`sync` issues for now. Wdyt?